### PR TITLE
Change resource key logic for k8s

### PR DESCRIPTION
**Pull request by ffjlabo**
**What this PR does / why we need it**:

This fix will solve the problem that the cluster-scoped resource can't be deleted when we set the namespace in app.pipecd.yaml.
context: https://github.com/pipe-cd/pipecd/issues/4269#issuecomment-2104055928

**expected behavior**
- We can delete the cluster-scoped resource.

**livestate side**
- use "" for the cluster-scoped resource
- use the namespace for the namespace scoped resource

**When reading the manifests from git** 
- use "" for the cluster-scoped resource
-  for the namespace scoped resource,
   - use `spec.input.namespace` in app.pipecd.yaml if it is set.
   - use the namespace on the manifest from git if it is set.
   - use `default` if  both of `spec.input.namespace` and the namespace are not set

---

Currently, we use the resource key to identify each k8s resource.
It is created by `MakeResourceKey`, and the rule is below.

**livestate side**
- use `default` when the resource obj doesn't have the namespace.
- use the namespace when the resource obj has the namespace.

**When reading the manifests from git** 
- use `spec.input.namespace` in app.pipecd.yaml if it is set.
- use the namespace on the manifest from git if it is set.
- use `default` when the namespace on the manifest is "".

This rule doesn't consider the cluster-scoped resource.
So for example, cluster-scoped resource don't have any namespace, but if we set the `spec.input.namespace` in the app.pipecd.yaml, it sets the value as the namespace to the resource key.

This causes the problem that can't prune the resource because the resource key can't identify the resource correctly.
So I fixed the the logic like below.

**livestate side**
- use "" for the cluster-scoped resource
- use the namespace for the namespace scoped resource

**When reading the manifests from git** 
- use "" for the cluster-scoped resource
-  for the namespace scoped resource,
   - use `spec.input.namespace` in app.pipecd.yaml if it is set.
   - use the namespace on the manifest from git if it is set.
   - use `default` if  both of `spec.input.namespace` and the namespace are not set

**Which issue(s) this PR fixes**:

Part of #4269

- **Does this PR introduce a user-facing change?**:

- **How are users affected by this change**:
- **Is this breaking change**:
- **How to migrate (if breaking change)**:

### CODE REVIEW

**Comment by ffjlabo** [comment]
# The Overview of the fix

**refine the namespace by `loader.refineNamespace` when reading the manifests from git repo**
- fixed to obtain information to determine whether each k8s resource is cluster scoped when executor, planner, plan preview, and detector start. (`provider.GetIsNamespacedResources`)

**remove the logic to determine the namespace from `MakeResourceKey`**
- I think `MakeResourceKey` shouldn't determine the namespace. This is  only loader's responsibility.

**Comment by ffjlabo** [comment]
@khanhtc1202 @t-kikuc 
This issue is complex and my explanation may be insufficient. If you have any questions, please feel free to ask. :) 🙏

**Comment by khanhtc1202** [comment]
@ffjlabo some testcases are failed :eyes:

**Comment by ffjlabo** [comment]
@khanhtc1202 Sorry, I fixed some of them. 🙏 
There are some failings, but these are under the /pipecdv1. I want to fix it after the current implementation looks good.

**Comment by khanhtc1202** [comment]
@ffjlabo Please rebase this PR due to pipedv1 package change 👀

**Comment by ffjlabo** [comment]
The failing test is solved by this PR 🙏 https://github.com/pipe-cd/pipecd/pull/4926

File: pkg/app/piped/executor/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -75,6 +77,18 @@ func (e *deployExecutor) Execute(sig executor.StopSignal) model.StageStatus {
 	ctx := sig.Context()
 	e.commit = e.Deployment.Trigger.Commit.Hash
 
+	cp, ok := e.PipedConfig.FindPlatformProvider(e.Deployment.PlatformProvider, model.ApplicationKind_KUBERNETES)
+	if !ok {
+		e.LogPersister.Errorf("provider %s was not found", e.Deployment.PlatformProvider)
+		return model.StageStatus_STAGE_FAILURE
+	}
```
Comment (khanhtc1202) [inline pkg/app/piped/executor/kubernetes/kubernetes.go:84]: Is this change related to PR purposes?

File: pkg/app/piped/executor/kubernetes/rollback.go [hunk 2]
```diff
@@ -41,6 +46,39 @@ func (e *rollbackExecutor) Execute(sig executor.StopSignal) model.StageStatus {
...
+	if err != nil {
+		e.LogPersister.Errorf("failed to create discovery client: %v", zap.Error(err))
+		return model.StageStatus_STAGE_FAILURE
+	}
+	groupResources, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		e.LogPersister.Errorf("failed to fetch preferred resources: %v", zap.Error(err))
+		return model.StageStatus_STAGE_FAILURE
+	}
+	e.LogPersister.Info(fmt.Sprintf("successfully preferred resources that contains for %d groups", len(groupResources)))
```
Comment (khanhtc1202) [inline pkg/app/piped/executor/kubernetes/rollback.go:72]:
```suggestion
	e.LogPersister.Infof("successfully preferred resources that contains for %d groups", len(groupResources))
```
And remove "fmt" package

File: pkg/app/piped/platformprovider/kubernetes/loader.go [hunk 3]
```diff
@@ -166,16 +167,50 @@ func (l *loader) LoadManifests(ctx context.Context) (manifests []Manifest, err e
...
-	if namespace == "" {
-		return
+// refineNamespace fix the namespace of the given manifest.
+// The priority is as follows:
+// If the resource is cluster-scoped, it returns an empty string.
+// Otherwise, it is the namespace-scoped resource and the namespace is determined by the following order:
+// 1. The namespace set in the application configuration.
+// 2. The namespace set in the manifest.
+// 3. The default namespace.
+func (l *loader) refineNamespace(m *Manifest) error {
```
Comment (khanhtc1202) [inline pkg/app/piped/platformprovider/kubernetes/loader.go:190]: Should it be named `determineResourceNamespace` or `determineNamespace`? 🤔

Comment (ffjlabo) [reply pkg/app/piped/platformprovider/kubernetes/loader.go]: Thanks! fixed [5d22c75](https://github.com/pipe-cd/pipecd/pull/4916/commits/5d22c7587fec159fb2a8e926ade63ded2eead06b)

File: pkg/app/piped/platformprovider/kubernetes/loader.go [hunk 4]
```diff
@@ -166,16 +167,50 @@ func (l *loader) LoadManifests(ctx context.Context) (manifests []Manifest, err e
...
+		if l.input.Namespace != "" {
+			namespace = l.input.Namespace
+		}
 	}
+
+	m.Key.Namespace = namespace
+	m.u.SetNamespace(namespace)
+
+	return nil
 }
```
Comment (ffjlabo) [inline pkg/app/piped/platformprovider/kubernetes/loader.go:214]: 📝 Check the current logic

**Comment by ffjlabo** [comment]
I tested it using QuickSync with the scenario below, and it works well.

scenario:
tried the following operations with and without setting `spec.input.namespace` in `app.pipecd.yaml` for cluster scoped resource and namespaced resource.
- create a k8s app
- update manifests
- fix the actual resource with `kubectl edit`
- delete manifests

I chose and tried them with ClusterRole (cluster scoped) and Deployment (namespaced) resources.

checkpoint:
- create, update, and delete the resources successfully
- reflect the operations result on the application livestate (check on the Web UI)
- detect OutOfSync after `kubectl edit`

File: pkg/app/piped/driftdetector/kubernetes/detector.go [hunk 5]
```diff
@@ -87,18 +89,19 @@ func NewDetector(
...
+		gitClient:             gitClient,
+		stateGetter:           stateGetter,
+		reporter:              reporter,
+		appManifestsCache:     appManifestsCache,
+		interval:              time.Minute,
+		config:                cfg,
+		secretDecrypter:       sd,
+		gitRepos:              make(map[string]git.Repo),
+		syncStates:            make(map[string]model.ApplicationSyncState),
+		isNamespacedResources: make(map[schema.GroupVersionKind]bool),
```
Comment (Warashi) [inline pkg/app/piped/driftdetector/kubernetes/detector.go:103]: nits

isNamespacedResources seems to be an immutable map. After construction, there is no operation to add an element to this map.
In this case, a nil map is enough to express an empty map.

example code
https://go.dev/play/p/A8qeO9dCyuw

**Review by Warashi — APPROVED** [review]
Seems good to me.
I commented on one nitpick for your information.

File: pkg/app/piped/driftdetector/kubernetes/detector.go (hunk 5 above)
Comment (ffjlabo) [reply pkg/app/piped/driftdetector/kubernetes/detector.go]: After that, I removed `isNamespacedResources` from the struct detector because it currently gets it every time it executes the detector's check logic. 🙏 
[a885bab](https://github.com/pipe-cd/pipecd/pull/4916/commits/a885bab8f238a6cf0dd224155eda5a556eeb3f25)

**Review by Warashi — APPROVED** [review]
LGTM

**Review by khanhtc1202 — APPROVED** [review]
Great work, thank you 🪨

**Review by t-kikuc — APPROVED** [review]
greatest work

**Comment by ffjlabo** [comment]
Thank you for reviewing 🙏
